aspnet/mvc/controllers/filters.rst

```diff
@@ -1,8 +1,327 @@
+
+Synchronous filters define both an On\ *Stage*\ Executing and On\ *Stage*\ Executed method (with noted exceptions). The On\ *Stage*\ Executing method will be called before the event pipeline stage by the Stage name, and the On\ *Stage*\ Executed method will be called after the pipeline stage named by the Stage name.
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/SampleActionFilter.cs
+  :language: c#
+  :emphasize-lines: 6,8,13
+
+Asynchronous filters define a single On\ *Stage*\ ExecutionAsync method that will surround execution of the pipeline stage named by Stage. The On\ *Stage*\ ExecutionAsync method is provided a *Stage*\ ExecutionDelegate delegate which will execute the pipeline stage named by Stage when invoked and awaited.
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/SampleAsyncActionFilter.cs
+  :language: c#
+  :emphasize-lines: 6,8-9
+
+.. note:: You should only implement *either* the synchronous or the async version of a filter interface, not both. If you need to perform async work in the filter, implement the async interface. Otherwise, implement the synchronous interface. The framework will check to see if the filter implements the async interface first, and if so, it will call it. If not, it will call the synchronous interface's method(s). If you were to implement both interfaces on one class, only the async method would be called by the framework. Also, it doesn't matter whether your action is async or not, your filters can be synchronous or async independent of the action.
+
+Filters and Attributes
+^^^^^^^^^^^^^^^^^^^^^^
+It's often convenient to implement filter interfaces as *Attributes*. Filter attributes are applied to controllers and action methods. The framework includes built-in attribute-based filters that you can subclass and customize. For example, the following filter inherits from `ResultFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/ResultFilterAttribute/index.html>`_, and overrides its ``OnResultExecuting`` method to add a header to the response.
+
+.. _add-header-attribute:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/AddHeaderAttribute.cs
+  :language: c#
+  :emphasize-lines: 5,16
+
+Attributes allow filters to accept arguments, as shown in the example above. You would add this attribute to a controller or action method and specify the name and value of the HTTP header you wished to add to the response:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Controllers/SampleController.cs
+  :language: c#
+  :emphasize-lines: 1
+  :lines: 6-12,25
+  :dedent: 4
+
+The result of the ``Index`` action is shown below - the response headers are displayed on the bottom right.
+
+.. image:: filters/_static/add-header.png
+
+Several of the filter interfaces have corresponding attributes that can be used as base classes for custom implementations.
+
+Filter attributes:
```

Comment:
Add a blank line before the bulleted list and don't indent the list. Otherwise this will render as a [definition list](http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#definition-lists)